### ISSUE TITLE
Can we use published packages instead of rebuilding?

### ISSUE THREAD

**Issue by bengry** [lerna/lerna#1983]
## Expected Behavior

Currently we have our repo set up such that both local development and CI do `lerna bootstrap`, which in turn runs the `prepare` script for all packages, which in turn builds each of them (`npm run build` -> `some-build-command-for-that-package` (e.g. `tsc`)).

## Current Behavior
This is fine, except for the fact that when we have lots of packages (currently ~35), this takes a lot of time to do (can be ~10 minutes). At the same time, we do publish the built packages post-build, and thus we already have them built in the npm feed.

## Possible Solution
When in CI - check if `foo@1.2.3` exists in the npm feed - use that, otherwise, build locally.

## Context
- Shortening CI build time

## Your Environment
- VSTS

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.13.1 |
| `npm --version`  | 6.9.0 |
| `node --version` | 10.x |

**Comment by bengry** [comment]
@evocateur Friendly ping on this, been a week with no response.

**Comment by evocateur** [comment]
Sorry, I've been really busy at my day job.

You should look into [Project References](https://www.typescriptlang.org/docs/handbook/project-references.html) if you haven't already, there's no _requirement_ to run `tsc` directly from each leaf package. If `tsc` is needlessly rebuilding already-built, unchanged code (unlike [pectin](https://github.com/evocateur/pectin#readme), for example), that's not a Lerna problem.

As for allowing released packages instead of local builds, I don't see that happening, as it breaks one of the fundamental contracts of a Lerna monorepo. One tradeoff of that is the builds will always run.

**Comment by bengry** [comment]
No worries.

Most of our projects inside the repo are indeed TypeScript based, some maybe using project references would indeed help. I'll have to see how it plays with Angular libraries, they do some magic there sometimes which tends breaks standard stuff unfortunately.

However, at any rate - the issue is at a higher level - the CI. Just like I can ask the CI to restore a cache of my npm packages (external ones, that is) (e.g. [CircleCI](https://circleci.com/docs/2.0/caching/#caching-libraries), but most other CIs offer the same thing) - I'd like to be able to ask lerna to restore the cache of the built result, if one is available (on the npm registry).

Even if TypeScript project references or the like would work - a CI usually starts from a clean state (via a container or otherwise), so incremental builds won't speed it up, only something at the Lerna-level would.

I'm not sure I understand why this "breaks one of the fundamental contracts of a Lerna monorepo", I'd appreciate further explanation of this. The requirement/feature request here is that if the same version of a package already exists in the npm feed exists - use that. If no such version exists, or if a different version is declared in the `package.json` - built it as being done today.

**Comment by benderr** [comment]
Hi. @bengry I ran into a similar problem and found a solution. Perhaps it will help you if you use the lerna in conjunction with the yarn workspaces. We used the `yarn install --focus` [command](https://yarnpkg.com/blog/2018/05/18/focused-workspaces/) to set the dependencies from the registry. The script in CI/CD looked something like this:
`
  yarn install //install all dependencies and symlinks
  lerna version ...params //bump version if there are changes in packages
  lerna publish ...params //publish packages
  cd packages/app //go to workspaces project "app" to build site
  yarn install --focus //install workspaces dependencies from registry (verdaccio in our case)
  yarn build //build site
`
